File: Heimdal/lib/gssapi/ntlm/acquire_cred.c
/*
 * Copyright (c) 1997 - 2004 Kungliga Tekniska Högskolan
 * (Royal Institute of Technology, Stockholm, Sweden).
 * All rights reserved.
 *
 * Portions Copyright (c) 2009 Apple Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the Institute nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "ntlm.h"
#include <gssapi_spi.h>

static OM_uint32
_gss_ntlm_have_cred(OM_uint32 *minor,
		    const ntlm_name target_name,
		    ntlm_cred *rcred)
{
    krb5_context context;
    krb5_error_code ret;
    krb5_storage *request, *response;
    krb5_data response_data;
    OM_uint32 major;
    ntlm_name cred;
    kcmuuid_t uuid;
    ssize_t sret;

    ret = krb5_init_context(&context);
    if (ret) {
	*minor = ret;
	return GSS_S_FAILURE;
    }

    ret = krb5_kcm_storage_request(context, KCM_OP_HAVE_NTLM_CRED, &request);
    if (ret)
	goto out;

    ret = krb5_store_stringz(request, target_name->user);
    if (ret)
	goto out;

    ret = krb5_store_stringz(request, target_name->domain);
    if (ret)
	goto out;

    ret = krb5_kcm_call(context, request, &response, &response_data);
    krb5_storage_free(request);
    if (ret)
	goto out;

    sret = krb5_storage_read(response, uuid, sizeof(uuid));

    krb5_storage_free(response);
    krb5_data_free(&response_data);

    if (sret != sizeof(uuid)) {
	krb5_clear_error_message(context);
	return KRB5_CC_IO;
    }

    major = _gss_ntlm_duplicate_name(minor, (gss_name_t)target_name,
				     (gss_name_t *)&cred);
    if (major)
	return major;

    cred->flags |= NTLM_UUID;
    memcpy(cred->uuid, uuid, sizeof(cred->uuid));

    *rcred = (ntlm_cred)cred;
 out:
    krb5_free_context(context);
    if (ret) {
	*minor = ret;
	major = GSS_S_FAILURE;
    }

    return major;
}

OM_uint32
_gss_ntlm_acquire_cred(OM_uint32 * min_stat,
		       const gss_name_t desired_name,
		       OM_uint32 time_req,
		       const gss_OID_set desired_mechs,
		       gss_cred_usage_t cred_usage,
		       gss_cred_id_t * output_cred_handle,
		       gss_OID_set * actual_mechs,
		       OM_uint32 * time_rec)
{
    ntlm_name name = (ntlm_name) desired_name;
    OM_uint32 maj_stat, junk;
    ntlm_ctx ctx;

    *min_stat = 0;
    *output_cred_handle = GSS_C_NO_CREDENTIAL;
    if (actual_mechs)
	*actual_mechs = GSS_C_NO_OID_SET;
    if (time_rec)
	*time_rec = GSS_C_INDEFINITE;

    if (desired_name == NULL)
	return GSS_S_NO_CRED;

    if (cred_usage == GSS_C_BOTH || cred_usage == GSS_C_ACCEPT) {
	gss_ctx_id_t gctx;

	maj_stat = _gss_ntlm_allocate_ctx(min_stat, name->domain, &ctx);
	if (maj_stat != GSS_S_COMPLETE)
	    return maj_stat;

	gctx = (gss_ctx_id_t)ctx;
	_gss_ntlm_delete_sec_context(&junk, &gctx, NULL);
    }	
    if (cred_usage == GSS_C_BOTH || cred_usage == GSS_C_INITIATE) {
	ntlm_cred cred;

	/* if we have a anon name, lets dup it directly */
	if ((name->flags & NTLM_ANON_NAME) != 0) {
	    maj_stat = _gss_ntlm_duplicate_name(min_stat,
						(gss_name_t)name,
						(gss_name_t *)&cred);
	    if (maj_stat)
		return maj_stat;
	} else {
	    maj_stat = _gss_ntlm_have_cred(min_stat, name, &cred);
	    if (maj_stat)
		return maj_stat;
	}
	*output_cred_handle = (gss_cred_id_t)cred;
    }

    return (GSS_S_COMPLETE);
}

OM_uint32
_gss_ntlm_acquire_cred_ext(OM_uint32 * minor_status,
			   const gss_name_t desired_name,
			   gss_const_OID credential_type,
			   const void *credential_data,
			   OM_uint32 time_req,
			   gss_const_OID desired_mech,
			   gss_cred_usage_t cred_usage,
			   gss_cred_id_t * output_cred_handle)
{
    ntlm_name name = (ntlm_name) desired_name;
    OM_uint32 major;
    krb5_storage *request, *response;
    krb5_data response_data;
    krb5_context context;
    krb5_error_code ret;
    struct ntlm_buf buf;
    krb5_data data;
    ntlm_cred dn;
    kcmuuid_t uuid;
    ssize_t sret;

    if (credential_data == NULL)
	return GSS_S_FAILURE;

    if (!gss_oid_equal(credential_type, GSS_C_CRED_PASSWORD))
	return GSS_S_FAILURE;

    if (name == NULL)
	return GSS_S_FAILURE;
	
    ret = krb5_init_context(&context);
    if (ret)
	return GSS_S_FAILURE;

    {
	gss_buffer_t buffer;
	char *password;
	
	buffer = (gss_buffer_t)credential_data;
	password = malloc(buffer->length + 1);
	if (password == NULL) {
	    ret = ENOMEM;
	    goto out;
	}
	memcpy(password, buffer->value, buffer->length);
	password[buffer->length] = '\0';
	
	heim_ntlm_nt_key(password, &buf);
	memset(password, 0, strlen(password));
	free(password);
    }

    data.data = buf.data;
    data.length = buf.length;
	
    krb5_kcm_storage_request(context, KCM_OP_ADD_NTLM_CRED, &request);
	
    krb5_store_stringz(request, name->user);
    krb5_store_stringz(request, name->domain);
    krb5_store_data(request, data);
    
    ret = krb5_kcm_call(context, request, &response, &response_data);
    krb5_storage_free(request);
    if (ret)
	goto out;
    
    sret = krb5_storage_read(response, &uuid, sizeof(uuid));

    krb5_storage_free(response);
    krb5_data_free(&response_data);

    if (sret != sizeof(uuid)) {
	ret = KRB5_CC_IO;
	goto out;
    }

    heim_ntlm_free_buf(&buf);

    dn = calloc(1, sizeof(*dn));
    if (dn == NULL) {
	major = GSS_S_FAILURE;
	goto out;
    }

    dn->user = strdup(name->user);
    dn->domain = strdup(name->domain);
    dn->flags = NTLM_UUID;
    memcpy(dn->uuid, uuid, sizeof(dn->uuid));

    *output_cred_handle = (gss_cred_id_t)dn;
    
    major = GSS_S_COMPLETE;
 out:

    krb5_free_context(context);
    if (ret)
	major = GSS_S_FAILURE;

    return major;
}

File: Heimdal/lib/heimcred/acquirecred.m
/*-
 * Copyright (c) 2013 Kungliga Tekniska Högskolan
 * (Royal Institute of Technology, Stockholm, Sweden).
 * All rights reserved.
 *
 * Portions Copyright (c) 2020 Apple Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#import "acquirecred.h"
#import "krb5.h"
#import "heimcred.h"
#import "gsscred.h"
#import "common.h"
#import "GSSCredXPCHelperClient.h"

void cred_update_expire_time_locked(HeimCredRef cred, time_t t);

static dispatch_queue_t event_work_queue;
static void init_event_queue() {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
	event_work_queue = dispatch_queue_create("com.apple.GSSCred.event_work", DISPATCH_QUEUE_SERIAL);
    });
}

//timer function used to renew renewable credentials
void
renew_func(heim_event_t event, void *ptr)
{
    init_event_queue();
    
    dispatch_sync(event_work_queue, ^{
	krb5_error_code ret;
	NSString *clientName;
	time_t expire;
	
	HeimCredRef cred = (HeimCredRef)ptr;
	if (cred==NULL) {
	    return;
	}
	heim_ipc_event_cancel(cred->renew_event);
	cred->next_acquire_time = 0;
	
	clientName = (__bridge NSString*)CFDictionaryGetValue(cred->attributes, kHEIMAttrClientName);
	os_log_debug(GSSOSLog(), "renew_func: %@", clientName);
	
	ret = [HeimCredGlobalCTX.gssCredHelperClientClass refreshForCred:cred expireTime:&expire];
	
	switch (ret) {
	    case KRB5KRB_AP_ERR_BAD_INTEGRITY:
	    case KRB5KRB_AP_ERR_MODIFIED:
	    case KRB5KDC_ERR_PREAUTH_FAILED:
		/* bad password, drop it like dead */
		os_log_debug(GSSOSLog(), "cache: %@ got auth failed, stop renewing",
			     clientName);
		HEIMDAL_MUTEX_lock(&cred->event_mutex);
		cred->acquire_status = CRED_STATUS_ACQUIRE_STOPPED;
		HEIMDAL_MUTEX_unlock(&cred->event_mutex);
		break;
	    case KRB5KRB_AP_ERR_TKT_EXPIRED:
		os_log_debug(GSSOSLog(), "The ticket has expired, stop renewing: %@",
			     clientName);
		HEIMDAL_MUTEX_lock(&cred->event_mutex);
		cred->acquire_status = CRED_STATUS_ACQUIRE_STOPPED;
		HEIMDAL_MUTEX_unlock(&cred->event_mutex);
		break;
	    case KRB5_CC_NOTFOUND:
		/* this is fatal, the requested cache is not found */
		os_log_debug(GSSOSLog(), "cache not found, stop renewing: %@",
			     clientName);
		HEIMDAL_MUTEX_lock(&cred->event_mutex);
		cred->acquire_status = CRED_STATUS_ACQUIRE_STOPPED;
		HEIMDAL_MUTEX_unlock(&cred->event_mutex);
		break;
	    case 0:
		os_log_debug(GSSOSLog(), "cache: %@ got new tickets (expire in %d seconds)",
			     clientName, (int)(expire - time(NULL)));
		//when the renewal is successful, this cred is replaced with another one.
		//updating these values does nothing
		break;
	    default:
		cred_update_renew_time(cred, true);
		HEIMDAL_MUTEX_lock(&cred->event_mutex);
		cred->acquire_status = CRED_STATUS_ACQUIRE_FAILED;
		HEIMDAL_MUTEX_unlock(&cred->event_mutex);
		break;
	}
    });
}

//event function used to acquire credentials or notify caches when a cred expires
void
expire_func(heim_event_t event, void *ptr)
{
    init_event_queue();
    
    dispatch_sync(event_work_queue, ^{
	krb5_error_code ret;
	NSString *clientName;
	time_t expire;
	
	HeimCredRef cred = (HeimCredRef)ptr;
	if (cred==NULL) {
	    return;
	}
	HEIMDAL_MUTEX_lock(&cred->event_mutex);
	heim_ipc_event_cancel(cred->renew_event);
	cred->next_acquire_time = 0;
	HEIMDAL_MUTEX_unlock(&cred->event_mutex);
	
	if (!isAcquireCred(cred)) {
	    if (cred->mech->notifyCaches!=NULL) {
		cred->mech->notifyCaches();
	    }
	    return;
	}
	
	clientName = (__bridge NSString*)CFDictionaryGetValue(cred->attributes, kHEIMAttrClientName);
	os_log_debug(GSSOSLog(), "expire_func: %@", clientName);
	
	ret = [HeimCredGlobalCTX.gssCredHelperClientClass acquireForCred:cred expireTime:&expire];
	
	switch (ret) {
	    case KRB5KRB_AP_ERR_BAD_INTEGRITY:
	    case KRB5KRB_AP_ERR_MODIFIED:
	    case KRB5KDC_ERR_PREAUTH_FAILED:
	    {
		/* bad password, drop it like dead */
		os_log_debug(GSSOSLog(), "cache: %@ got bad password, stop renewing",
			     clientName);
		cred_update_acquire_status(cred, CRED_STATUS_ACQUIRE_STOPPED);
	    }
		break;
	    case KRB5_CC_NOTFOUND:
	    {
		/* this is fatal, the requested cache is not found */
		os_log_debug(GSSOSLog(), "cache not found, stop renewing: %@",
			     clientName);
		cred_update_acquire_status(cred, CRED_STATUS_ACQUIRE_STOPPED);
	    }
		break;
	    case 0:
	    {
		os_log_debug(GSSOSLog(), "cache: %@ got new tickets (expire in %d seconds)",
			     clientName, (int)(expire - time(NULL)));
		HEIMDAL_MUTEX_lock(&cred->event_mutex);
		cred->expire = expire;
		HEIMDAL_MUTEX_unlock(&cred->event_mutex);
		cred_update_acquire_status(cred, CRED_STATUS_ACQUIRE_SUCCESS);
	    }
		break;
	    default:
		HEIMDAL_MUTEX_lock(&cred->event_mutex);
		cred_update_expire_time_locked(cred, time(NULL) + 300);
		HEIMDAL_MUTEX_unlock(&cred->event_mutex);
		cred_update_acquire_status(cred, CRED_STATUS_ACQUIRE_FAILED);
		break;
	}

	HeimCredGlobalCTX.executeOnRunQueue(^{
	    HeimCredCTX.needFlush = 1;
	    HeimCredGlobalCTX.saveToDiskIfNeeded();
	});

    });
}

void
final_func(void *ptr)
{
}

#define KRB5_CONF_NAME "krb5_ccache_conf_data"
#define KRB5_REALM_NAME "X-CACHECONF:"

void
cred_update_acquire_status(HeimCredRef cred, int status)
{
    os_log_debug(GSSOSLog(), "cred_update_acquire_status: %@", CFBridgingRelease(CFUUIDCreateString(NULL, cred->uuid)));
    
    HEIMDAL_MUTEX_lock(&cred->event_mutex);
    cred->acquire_status = status;
    
    switch (status) {
	case CRED_STATUS_ACQUIRE_START:
	    cred_update_expire_time_locked(cred, time(NULL) + 2);  //wait a few seconds to start to let load and save methods finish.
	    break;

	case CRED_STATUS_ACQUIRE_STOPPED:
	    cred->next_acquire_time = 0;
	    heim_ipc_event_cancel(cred->expire_event);
	    break;

	case CRED_STATUS_ACQUIRE_FAILED:
	    cred_update_expire_time_locked(cred, time(NULL) + 300);
	    break;

	case CRED_STATUS_ACQUIRE_SUCCESS: {
	    time_t next_refresh, now = time(NULL);

	    if (cred->expire > now) {
		next_refresh = cred->expire;
		/* try to acquire credential just before */
		if (cred->expire - now > 300)
		    next_refresh -= 300;
		cred_update_expire_time_locked(cred, next_refresh);
	    } else {
		cred->next_acquire_time = 0;
	    }
	    break;
	}
    }
    HEIMDAL_MUTEX_unlock(&cred->event_mutex);
    
}

//call this in the event mutex
void
cred_update_expire_time_locked(HeimCredRef cred, time_t t)
{
    if (t == 0) {
	t = time(NULL);
    } else if (t < time(NULL)) {
	cred->next_acquire_time = 0;
	os_log_debug(GSSOSLog(), "%@: acquire time is in the past",
		     CFBridgingRelease(CFUUIDCreateString(NULL, cred->uuid)));
	return;
    }
    cred->next_acquire_time = t;
    heim_ipc_event_set_time(cred->expire_event, t);
    os_log_debug(GSSOSLog(), "%@: will try to acquire credentals in %d seconds",
		 CFBridgingRelease(CFUUIDCreateString(NULL, cred->uuid)), (int)(t - time(NULL)));
}

#define KCM_EVENT_QUEUE_INTERVAL 60

void
cred_update_renew_time(HeimCredRef cred, bool is_retry)
{
    time_t renewtime = time(NULL) + HeimCredGlobalCTX.renewInterval;
    
    if (is_retry) {
	renewtime = time(NULL) + 300;
    }
    HEIMDAL_MUTEX_lock(&cred->event_mutex);
    time_t expire = cred->expire;
    
    /* if the ticket is about to expire in less then QUEUE_INTERVAL,
     * don't bother */
    if (time(NULL) + KCM_EVENT_QUEUE_INTERVAL > expire) {
	os_log_debug(GSSOSLog(), "%@: will expire before next attempt",
		     CFBridgingRelease(CFUUIDCreateString(NULL, cred->uuid)));
	HEIMDAL_MUTEX_unlock(&cred->event_mutex);
	return;
    }
    
    if (renewtime > expire - KCM_EVENT_QUEUE_INTERVAL)
	renewtime = expire - KCM_EVENT_QUEUE_INTERVAL;
    
    os_log_debug(GSSOSLog(), "%@: will try to renew credentals in %d seconds",
	    cred, (int)(renewtime - time(NULL)));
    
    heim_ipc_event_set_time(cred->renew_event, renewtime);
    cred->renew_time = renewtime;
    HEIMDAL_MUTEX_unlock(&cred->event_mutex);
}
